File: ios/Classes/Entities.swift
import PassKit

struct Constants {
    static let paymentErrorContactInvalid = "paymentErrorContactInvalid"
    static let paymentErrorShippingAddressInvalid = "paymentErrorShippingAddressInvalid"
    static let paymentErrorBillingAddressInvalid = "paymentErrorBillingAddressInvalid"
    static let paymentErrorShippingAddressUnserviceable = "paymentErrorShippingAddressUnserviceable"

    static let paymentResultShippingContactUpdate = "paymentResultShippingContactUpdate"
    static let paymentResultShippingMethodUpdate = "paymentResultShippingMethodUpdate"
    static let paymentResultPaymentMethodUpdate = "paymentResultPaymentMethodUpdate"
}

struct APayPaymentAuthStatus: Codable {
    let value: Int

    static let success = APayPaymentAuthStatus(value: 0)
    static let failure = APayPaymentAuthStatus(value: 1)

    public func toPK() -> PKPaymentAuthorizationStatus {
        PKPaymentAuthorizationStatus(rawValue: value)!
    }
}

struct APayPaymentMethodType: Codable {
    let value: Int

    static let unknown = APayPaymentMethodType(value: 0)
    static let debit = APayPaymentMethodType(value: 1)
    static let credit = APayPaymentMethodType(value: 2)
    static let prepaid = APayPaymentMethodType(value: 3)
    static let store = APayPaymentMethodType(value: 4)

    static func fromPK(_ value: PKPaymentMethodType) -> APayPaymentMethodType {
        APayPaymentMethodType(value: Int(value.rawValue))
    }
}

struct APayPaymentSummaryItemType: Codable {
    let value: Int

    static let final = APayPaymentSummaryItemType(value: 0)
    static let pending = APayPaymentSummaryItemType(value: 1)

    public static func fromPK(_ value: PKPaymentSummaryItemType) -> APayPaymentSummaryItemType {
        APayPaymentSummaryItemType(value: Int(value.rawValue))
    }

    public func toPK() -> PKPaymentSummaryItemType? {
        PKPaymentSummaryItemType(rawValue: UInt(value))
    }
}

struct APayShippingType: Codable {
    let value: String

    static let shipping = APayShippingType(value: "shipping")
    static let delivery = APayShippingType(value: "delivery")
    static let storePickup = APayShippingType(value: "storePickup")
    static let servicePickup = APayShippingType(value: "servicePickup")

    public func toPK() -> PKShippingType? {
        switch (value) {
        case "shipping": return .shipping
        case "delivery": return .delivery
        case "storePickup": return .storePickup
        case "servicePickup": return .servicePickup
        default: return nil
        }
    }
}

struct APayMerchantCapability: Codable {
    let value: String

    static let capability3DS = APayMerchantCapability(value: "capability3DS")
    static let capabilityEMV = APayMerchantCapability(value: "capabilityEMV")
    static let capabilityCredit = APayMerchantCapability(value: "capabilityCredit")
    static let capabilityDebit = APayMerchantCapability(value: "capabilityDebit")

    public func toPK() -> PKMerchantCapability? {
        switch (value) {
        case "capability3DS": return .capability3DS
        case "capabilityEMV": return .capabilityEMV
        case "capabilityCredit": return .capabilityCredit
        case "capabilityDebit": return .capabilityDebit
        default: return nil
        }
    }
}

struct APayPaymentNetwork: Codable {
    let value: String

    static let carteBancaire = APayPaymentNetwork(value: "carteBancaire")
    static let carteBancaires = APayPaymentNetwork(value: "carteBancaires")
    static let cartesBancaires = APayPaymentNetwork(value: "cartesBancaires")
    static let chinaUnionPay = APayPaymentNetwork(value: "chinaUnionPay")
    static let discover = APayPaymentNetwork(value: "discover")
    static let eftpos = APayPaymentNetwork(value: "eftpos")
    static let electron = APayPaymentNetwork(value: "electron")
    static let elo = APayPaymentNetwork(value: "elo")
    static let idCredit = APayPaymentNetwork(value: "idCredit")
    static let interac = APayPaymentNetwork(value: "interac")
    static let JCB = APayPaymentNetwork(value: "JCB")
    static let mada = APayPaymentNetwork(value: "mada")
    static let maestro = APayPaymentNetwork(value: "maestro")
    static let masterCard = APayPaymentNetwork(value: "masterCard")
    static let privateLabel = APayPaymentNetwork(value: "privateLabel")
    static let quicPay = APayPaymentNetwork(value: "quicPay")
    static let suica = APayPaymentNetwork(value: "suica")
    static let visa = APayPaymentNetwork(value: "visa")
    static let vPay = APayPaymentNetwork(value: "vPay")
    static let barcode = APayPaymentNetwork(value: "barcode")
    static let girocard = APayPaymentNetwork(value: "girocard")

    public func toPK() -> PKPaymentNetwork? {
        switch value {
        case "carteBancaire": return .carteBancaire
        case "carteBancaires": return .carteBancaires
        case "cartesBancaires": return .cartesBancaires
        case "chinaUnionPay": return .chinaUnionPay
        case "discover": return .discover
        case "eftpos": return .eftpos
        case "electron": return .electron
        case "elo":
            if #available(iOS 12.1.1, *) {
                return .elo
            }
            return nil
        case "idCredit": return .idCredit
        case "interac": return .interac
        case "JCB": return .JCB
        case "mada":
            if #available(iOS 12.1.1, *) {
                return .mada
            }
            return nil
        case "maestro": return .maestro
        case "masterCard": return .masterCard
        case "privateLabel": return .privateLabel
        case "quicPay": return .quicPay
        case "suica": return .suica
        case "visa": return .visa
        case "vPay": return .vPay
        case "barcode":
            if #available(iOS 14, *) {
                return .barcode
            }
            return nil
        case "girocard":
            if #available(iOS 14, *) {
                return .girocard
            }
            return nil
        default: return nil
        }
    }

    static func fromPK(_ value: PKPaymentNetwork) -> APayPaymentNetwork? {
        switch (value) {
        case .carteBancaire: return carteBancaire
        case .carteBancaires: return carteBancaires
        case .cartesBancaires: return cartesBancaires
        case .chinaUnionPay: return chinaUnionPay
        case .discover: return discover
        case .eftpos: return eftpos
        case .electron: return electron
        case .idCredit: return idCredit
        case .interac: return interac
        case .JCB: return JCB
        case .maestro: return maestro
        case .masterCard: return masterCard
        case .privateLabel:return privateLabel
        case .quicPay: return quicPay
        case .suica:return suica;
        case .visa: return visa
        case .vPay: return vPay
        default:
            if #available(iOS 12.1.1, *) {
                if value == .elo {
                    return elo
                }
                if value == .mada {
                    return mada
                }
            }
            if #available(iOS 14, *) {
                if value == .barcode {
                    return barcode
                }
                if value == .girocard {
                    return girocard
                }
            }
            return nil
        }
    }
}

struct APayContactField: Codable {
    let value: String

    static let postalAddress = APayContactField(value: "postalAddress")
    static let emailAddress = APayContactField(value: "emailAddress")
    static let phoneNumber = APayContactField(value: "phoneNumber")
    static let name = APayContactField(value: "name")
    static let phoneticName = APayContactField(value: "phoneticName")

    public func toPK() -> PKContactField? {
        switch (value) {
        case "postalAddress": return .postalAddress
        case "emailAddress": return .emailAddress
        case "phoneNumber": return .phoneNumber
        case "name": return .name
        case "phoneticName": return .phoneticName
        default: return nil
        }
    }
}

struct APayPaymentSummaryItem: Codable {
    let label: String
    let amount: Double
    let type: APayPaymentSummaryItemType

    public func toPK() -> PKPaymentSummaryItem? {
        guard let paymentSummaryItemType = type.toPK() else {
            return nil
        }

        return PKPaymentSummaryItem(label: label, amount: NSDecimalNumber(value: amount), type: paymentSummaryItemType)
    }
}

struct APayShippingMethod: Codable {
    let label: String
    let amount: Double
    let type: APayPaymentSummaryItemType
    let identifier: String?
    let detail: String?

    public static func fromPK(_ value: PKShippingMethod) -> APayShippingMethod {
        APayShippingMethod(
                label: value.label,
                amount: value.amount.doubleValue,
                type: APayPaymentSummaryItemType.fromPK(value.type),
                identifier: value.identifier,
                detail: value.detail
        )
    }

    public func toPK() -> PKShippingMethod? {
        guard let paymentSummaryItemType = type.toPK() else {
            return nil
        }

        let result = PKShippingMethod(label: label, amount: NSDecimalNumber(value: amount), type: paymentSummaryItemType)

        result.identifier = identifier
        result.detail = detail

        return result
    }
}

class APayPersonNameComponents: Codable {
    public init(namePrefix: String?, givenName: String?, middleName: String?, familyName: String?, nameSuffix: String?, nickname: String?, phoneticRepresentation: APayPersonNameComponents?) {
        self.namePrefix = namePrefix
        self.givenName = givenName
        self.middleName = middleName
        self.familyName = familyName
        self.nameSuffix = nameSuffix
        self.nickname = nickname
        self.phoneticRepresentation = phoneticRepresentation
    }

    let namePrefix: String?
    let givenName: String?
    let middleName: String?
    let familyName: String?
    let nameSuffix: String?
    let nickname: String?
    let phoneticRepresentation: APayPersonNameComponents?

    public static func fromPK(_ value: PersonNameComponents) -> APayPersonNameComponents {
        APayPersonNameComponents(
                namePrefix: value.nameSuffix,
                givenName: value.givenName,
                middleName: value.middleName,
                familyName: value.familyName,
                nameSuffix: value.nameSuffix,
                nickname: value.nickname,
                phoneticRepresentation: value.phoneticRepresentation != nil ? APayPersonNameComponents.fromPK(value.phoneticRepresentation!) : nil
        )
    }

    public func toPK() -> PersonNameComponents {
        var result = PersonNameComponents()

        result.namePrefix = namePrefix
        result.givenName = givenName
        result.middleName = middleName
        result.familyName = familyName
        result.nameSuffix = nameSuffix
        result.nickname = nickname
        result.phoneticRepresentation = phoneticRepresentation?.toPK()

        return result
    }

}

struct APayPostalAddress: Codable {
    let street: String
    let subLocality: String
    let city: String
    let subAdministrativeArea: String
    let state: String
    let postalCode: String
    let country: String
    let isoCountryCode: String

    public static func fromPK(_ value: CNPostalAddress) -> APayPostalAddress {
        APayPostalAddress(
                street: value.street,
                subLocality: value.subLocality,
                city: value.city,
                subAdministrativeArea: value.subAdministrativeArea,
                state: value.state,
                postalCode: value.postalCode,
                country: value.country,
                isoCountryCode: value.isoCountryCode
        )
    }

    public func toPK() -> CNPostalAddress {
        let result = CNMutablePostalAddress()

        result.street = street
        result.subLocality = subLocality
        result.city = city
        result.subAdministrativeArea = subAdministrativeArea
        result.state = state
        result.postalCode = postalCode
        result.country = country
        result.isoCountryCode = isoCountryCode

        return result
    }
}

struct APayContact: Codable {
    let name: APayPersonNameComponents?
    let postalAddress: APayPostalAddress?
    let phoneNumber: String?
    let emailAddress: String?

    public static func fromPK(_ value: PKContact) -> APayContact {
        APayContact(
                name: value.name != nil ? APayPersonNameComponents.fromPK(value.name!) : nil,
                postalAddress: value.postalAddress != nil ? APayPostalAddress.fromPK(value.postalAddress!) : nil,
                phoneNumber: value.phoneNumber?.stringValue,
                emailAddress: value.emailAddress
        )
    }

    public func toPK() -> PKContact {
        let result = PKContact()

        result.name = name?.toPK()
        result.postalAddress = postalAddress?.toPK()
        result.phoneNumber = phoneNumber != nil ? CNPhoneNumber(stringValue: phoneNumber!) : nil
        result.emailAddress = emailAddress

        return result
    }
}

struct APayPaymentMethod: Codable {
    let displayName: String?
    let network: APayPaymentNetwork?
    let type: APayPaymentMethodType

    static func fromPK(_ value: PKPaymentMethod) -> APayPaymentMethod {
        APayPaymentMethod(
                displayName: value.displayName,
                network: value.network != nil ? APayPaymentNetwork.fromPK(value.network!) : nil,
                type: APayPaymentMethodType.fromPK(value.type)
        )
    }
}

struct APayPaymentToken: Codable {
    let paymentMethod: APayPaymentMethod
    let transactionIdentifier: String
    let paymentData: String

    public static func fromPK(_ value: PKPaymentToken) -> APayPaymentToken {
        APayPaymentToken(
                paymentMethod: APayPaymentMethod.fromPK(value.paymentMethod),
                transactionIdentifier: value.transactionIdentifier,
                paymentData: String(data: value.paymentData, encoding: .utf8)!
        )
    }
}

struct APayPayment: Codable {
    let token: APayPaymentToken
    let billingContact: APayContact?
    let shippingContact: APayContact?
    let shippingMethod: APayShippingMethod?

    public static func fromPK(_ value: PKPayment) -> APayPayment {
        APayPayment(
                token: APayPaymentToken.fromPK(value.token),
                billingContact: value.billingContact != nil ? APayContact.fromPK(value.billingContact!) : nil,
                shippingContact: value.shippingContact != nil ? APayContact.fromPK(value.shippingContact!) : nil,
                shippingMethod: value.shippingMethod != nil ? APayShippingMethod.fromPK(value.shippingMethod!) : nil
        )
    }
}

struct CanMakeRequest: Codable {
    let usingNetworks: [APayPaymentNetwork]?
    let capabilities: [APayMerchantCapability]?
}

struct APayPaymentError: Codable {
    let errorType: String
    let localizedDescription: String?
    let postalAddressKey: String?
    let contactField: APayContactField?

    public func toPK() -> Error? {
        switch (errorType) {
        case Constants.paymentErrorContactInvalid:
            guard let contactField = contactField?.toPK() else {
                return nil
            }

            return PKPaymentRequest.paymentContactInvalidError(withContactField: contactField, localizedDescription: localizedDescription)
        case Constants.paymentErrorShippingAddressInvalid:
            return PKPaymentRequest.paymentShippingAddressInvalidError(withKey: postalAddressKey!, localizedDescription: localizedDescription)
        case Constants.paymentErrorBillingAddressInvalid:
            return PKPaymentRequest.paymentBillingAddressInvalidError(withKey: postalAddressKey!, localizedDescription: localizedDescription)
        case Constants.paymentErrorShippingAddressUnserviceable:
            return PKPaymentRequest.paymentShippingAddressUnserviceableError(withLocalizedDescription: localizedDescription)
        default: return nil
        }
    }
}

struct ProcessPaymentRequest: Codable {
    let merchantIdentifier: String
    let countryCode: String
    let currencyCode: String
    let supportedNetworks: [APayPaymentNetwork]
    let merchantCapabilities: [APayMerchantCapability]
    let paymentSummaryItems: [APayPaymentSummaryItem]
    let requiredBillingContactFields: [APayContactField]?
    let requiredShippingContactFields: [APayContactField]?
    let billingContact: APayContact?
    let shippingContact: APayContact?
    let shippingMethods: [APayShippingMethod]?
    let shippingType: APayShippingType
    let applicationData: String?
    let supportedCountries: [String]?
}

/*
struct APayPaymentRequestUpdate {
    let status: APayPaymentAuthStatus
    let paymentSummaryItems: [APayPaymentSummaryItem]
}
*/

/// AuthorizePayment
struct APayPaymentAuthorizationResult: Codable {
    let status: APayPaymentAuthStatus
    let errors: [APayPaymentError]?

    public func toPK() -> PKPaymentAuthorizationResult {
        PKPaymentAuthorizationResult(
                status: status.toPK(),
                errors: errors?.map({ $0.toPK() }).onlyType()
        )
    }
}

struct AuthorizePaymentRequest: Codable {
    let id: Int
    let payment: APayPayment
}

struct AuthorizePaymentResponse: Codable {
    let result: APayPaymentAuthorizationResult
}

/// SelectShippingMethod
struct APayRequestShippingMethodUpdate: Codable {
    let status: APayPaymentAuthStatus
    let paymentSummaryItems: [APayPaymentSummaryItem]

    public func toPK() -> PKPaymentRequestShippingMethodUpdate {
        let result = PKPaymentRequestShippingMethodUpdate(
                paymentSummaryItems: paymentSummaryItems.map({ $0.toPK() }).onlyType()
        )
        result.status = status.toPK()
        return result
    }
}

struct SelectShippingMethodRequest: Codable {
    let id: Int
    let shippingMethod: APayShippingMethod
}

struct SelectShippingMethodResponse: Codable {
    let result: APayRequestShippingMethodUpdate
}

/// SelectShippingContact
struct APayRequestShippingContactUpdate: Codable {
    let status: APayPaymentAuthStatus
    let paymentSummaryItems: [APayPaymentSummaryItem]
    let shippingMethods: [APayShippingMethod]
    let errors: [APayPaymentError]?

    public func toPK() -> PKPaymentRequestShippingContactUpdate {
        let result = PKPaymentRequestShippingContactUpdate(
                errors: errors?.map({ $0.toPK() }).onlyType(),
                paymentSummaryItems: paymentSummaryItems.map({ $0.toPK() }).onlyType(),
                shippingMethods: shippingMethods.map({ $0.toPK() }).onlyType()
        )
        result.status = status.toPK()
        return result
    }
}

struct SelectShippingContactRequest: Codable {
    let id: Int
    let shippingContact: APayContact
}

struct SelectShippingContactResponse: Codable {
    let result: APayRequestShippingContactUpdate
}

/// SelectPaymentMethod
struct APayRequestPaymentMethodUpdate: Codable {
    let status: APayPaymentAuthStatus
    let paymentSummaryItems: [APayPaymentSummaryItem]
    let errors: [APayPaymentError]?

    public func toPK() -> PKPaymentRequestPaymentMethodUpdate {
        let result = PKPaymentRequestPaymentMethodUpdate(
                errors: errors?.map({ $0.toPK() }).onlyType(),
                paymentSummaryItems: paymentSummaryItems.map({ $0.toPK() }).onlyType()
        )
        result.status = status.toPK()
        return result
    }
}

struct SelectPaymentMethodRequest: Codable {
    let id: Int
    let paymentMethod: APayPaymentMethod
}

struct SelectPaymentMethodResponse: Decodable {
    let result: APayRequestPaymentMethodUpdate
}

File: lib/src/entities.dart
class PKPaymentAuthStatus {
  final int value;

  const PKPaymentAuthStatus(this.value);

  static PKPaymentAuthStatus fromJson(Map map) => PKPaymentAuthStatus(map['value'] as int);

  Map toJson() => {'value': value};

  static const success = PKPaymentAuthStatus(0);
  static const failure = PKPaymentAuthStatus(1);
}

class PKPaymentNetwork {
  final String value;

  const PKPaymentNetwork(this.value);

  static PKPaymentNetwork fromJson(Map map) => PKPaymentNetwork(map['value'] as String);

  Map toJson() => {'value': value};

  static const carteBancaire = PKPaymentNetwork("carteBancaire");
  static const carteBancaires = PKPaymentNetwork("carteBancaires");
  static const cartesBancaires = PKPaymentNetwork("cartesBancaires");
  static const chinaUnionPay = PKPaymentNetwork("chinaUnionPay");
  static const discover = PKPaymentNetwork("discover");
  static const eftpos = PKPaymentNetwork("eftpos");
  static const electron = PKPaymentNetwork("electron");
  static const elo = PKPaymentNetwork("elo");
  static const idCredit = PKPaymentNetwork("idCredit");
  static const interac = PKPaymentNetwork("interac");
  static const JCB = PKPaymentNetwork("JCB");
  static const mada = PKPaymentNetwork("mada");
  static const maestro = PKPaymentNetwork("maestro");
  static const masterCard = PKPaymentNetwork("masterCard");
  static const privateLabel = PKPaymentNetwork("privateLabel");
  static const quicPay = PKPaymentNetwork("quicPay");
  static const suica = PKPaymentNetwork("suica");
  static const visa = PKPaymentNetwork("visa");
  static const vPay = PKPaymentNetwork("vPay");
  static const barcode = PKPaymentNetwork("barcode");
  static const girocard = PKPaymentNetwork("girocard");
}

class PKMerchantCapability {
  final String value;

  const PKMerchantCapability(this.value);

  static PKMerchantCapability fromJson(Map map) => PKMerchantCapability(map['value'] as String);

  Map toJson() => {'value': value};

  static const capability3DS = PKMerchantCapability("capability3DS");
  static const capabilityEMV = PKMerchantCapability("capabilityEMV");
  static const capabilityCredit = PKMerchantCapability("capabilityCredit");
  static const capabilityDebit = PKMerchantCapability("capabilityDebit");
}

class PKShippingType {
  final String value;

  const PKShippingType(this.value);

  static PKShippingType fromJson(Map map) => PKShippingType(map['value'] as String);

  Map toJson() => {'value': value};

  static const shipping = PKShippingType("shipping");
  static const delivery = PKShippingType("delivery");
  static const storePickup = PKShippingType("storePickup");
  static const servicePickup = PKShippingType("servicePickup");
}

class PKContactField {
  final String value;

  const PKContactField(this.value);

  static PKContactField fromJson(Map map) => PKContactField(map['value'] as String);

  Map toJson() => {'value': value};

  static const postalAddress = PKContactField("postalAddress");
  static const emailAddress = PKContactField("emailAddress");
  static const phoneNumber = PKContactField("phoneNumber");
  static const name = PKContactField("name");
  static const phoneticName = PKContactField("phoneticName");

  static const List<PKContactField> all = [postalAddress, emailAddress, phoneNumber, name, phoneticName];
}

class PKPaymentSummaryItemType {
  final int value;

  const PKPaymentSummaryItemType(this.value);

  static PKPaymentSummaryItemType fromJson(Map map) => PKPaymentSummaryItemType(map['value'] as int);

  Map toJson() => {'value': value};

  static const finall = PKPaymentSummaryItemType(0);
  static const pending = PKPaymentSummaryItemType(1);
}

class PKPaymentMethodType {
  final int value;

  const PKPaymentMethodType(this.value);

  static PKPaymentMethodType fromJson(Map map) => PKPaymentMethodType(map['value'] as int);

  Map toJson() => {'value': value};

  static const unknown = PKPaymentSummaryItemType(0);
  static const debit = PKPaymentSummaryItemType(1);
  static const credit = PKPaymentSummaryItemType(0);
  static const prepaid = PKPaymentSummaryItemType(1);
  static const store = PKPaymentSummaryItemType(0);
}

class PKPaymentMethod {
  final String? displayName;
  final PKPaymentNetwork? network;
  final PKPaymentMethodType type;

  const PKPaymentMethod({this.displayName, this.network, required this.type});

  static PKPaymentMethod fromJson(Map map) => PKPaymentMethod(
        displayName: map['displayName'] as String?,
        network: map['network'] != null ? PKPaymentNetwork.fromJson(map['network'] as Map) : null,
        type: PKPaymentMethodType.fromJson(map['type'] as Map),
      );
}

class PKPaymentSummaryItem {
  final String label;
  final double amount;
  final PKPaymentSummaryItemType type;

  const PKPaymentSummaryItem({
    required this.label,
    required this.amount,
    this.type = PKPaymentSummaryItemType.finall,
  });

  static PKPaymentSummaryItem fromJson(Map map) {
    return PKPaymentSummaryItem(
      label: map['label'] as String,
      amount: (map['amount'] as num).toDouble(),
      type: PKPaymentSummaryItemType.fromJson(map['type'] as Map),
    );
  }

  Map toJson() => {'label': label, 'amount': amount, 'type': type.toJson()};
}

class PKShippingMethod {
  final String label;
  final double amount;
  final PKPaymentSummaryItemType type;
  final String? identifier;
  final String? detail;

  const PKShippingMethod({
    required this.label,
    required this.amount,
    this.type = PKPaymentSummaryItemType.finall,
    this.identifier,
    this.detail,
  });

  static PKShippingMethod fromJson(Map map) {
    return PKShippingMethod(
      label: map['label'] as String,
      amount: (map['amount'] as num).toDouble(),
      type: PKPaymentSummaryItemType.fromJson(map['type'] as Map),
      identifier: map['identifier'] as String?,
      detail: map['detail'] as String?,
    );
  }

  Map toJson() => {
        'label': label,
        'amount': amount,
        'type': type.toJson(),
        'identifier': identifier,
        'detail': detail,
      };
}

class PersonNameComponents {
  final String? namePrefix;
  final String? givenName;
  final String? middleName;
  final String? familyName;
  final String? nameSuffix;
  final String? nickname;
  final PersonNameComponents? phoneticRepresentation;

  const PersonNameComponents({
    this.namePrefix,
    this.givenName,
    this.middleName,
    this.familyName,
    this.nameSuffix,
    this.nickname,
    this.phoneticRepresentation,
  });

  static PersonNameComponents fromJson(Map map) {
    return PersonNameComponents(
      namePrefix: map['namePrefix'] as String?,
      givenName: map['givenName'] as String?,
      middleName: map['middleName'] as String?,
      familyName: map['familyName'] as String?,
      nameSuffix: map['nameSuffix'] as String?,
      nickname: map['nickname'] as String?,
      phoneticRepresentation: map['phoneticRepresentation'] != null
          ? PersonNameComponents.fromJson(map['phoneticRepresentation'] as Map)
          : null,
    );
  }

  Map toJson() => {
        'namePrefix': namePrefix,
        'givenName': givenName,
        'middleName': middleName,
        'familyName': familyName,
        'nameSuffix': nameSuffix,
        'nickname': nickname,
        'phoneticRepresentation': phoneticRepresentation?.toJson(),
      };
}

class CNPostalAddress {
  final String street;
  final String subLocality;
  final String city;
  final String subAdministrativeArea;
  final String state;
  final String postalCode;
  final String country;
  final String isoCountryCode;

  const CNPostalAddress({
    required this.street,
    required this.subLocality,
    required this.city,
    required this.subAdministrativeArea,
    required this.state,
    required this.postalCode,
    required this.country,
    required this.isoCountryCode,
  });

  static CNPostalAddress fromJson(Map map) {
    return CNPostalAddress(
      street: map['street'] as String,
      subLocality: map['subLocality'] as String,
      city: map['city'] as String,
      subAdministrativeArea: map['subAdministrativeArea'] as String,
      state: map['state'] as String,
      postalCode: map['postalCode'] as String,
      country: map['country'] as String,
      isoCountryCode: map['isoCountryCode'] as String,
    );
  }

  Map toJson() => {
        'street': street,
        'subLocality': subLocality,
        'city': city,
        'subAdministrativeArea': subAdministrativeArea,
        'state': state,
        'postalCode': postalCode,
        'country': country,
        'isoCountryCode': isoCountryCode,
      };
}

class PKContact {
  final PersonNameComponents? name;
  final CNPostalAddress? postalAddress;
  final String? phoneNumber;
  final String? emailAddress;

  const PKContact({this.name, this.postalAddress, this.phoneNumber, this.emailAddress});

  static PKContact fromJson(Map map) {
    return PKContact(
      name: map['name'] != null ? PersonNameComponents.fromJson(map['name'] as Map) : null,
      postalAddress: map['postalAddress'] != null ? CNPostalAddress.fromJson(map['postalAddress'] as Map) : null,
      phoneNumber: map['phoneNumber'] as String?,
      emailAddress: map['emailAddress'] as String?,
    );
  }

  Map toJson() => {
        'name': name?.toJson(),
        'postalAddress': postalAddress?.toJson(),
        'phoneNumber': phoneNumber,
        'emailAddress': emailAddress,
      };
}

class PKPaymentToken {
  final PKPaymentMethod paymentMethod;
  final String transactionIdentifier;
  final String paymentData;

  const PKPaymentToken({required this.paymentMethod, required this.transactionIdentifier, required this.paymentData});

  static PKPaymentToken fromJson(Map map) {
    return PKPaymentToken(
      paymentMethod: PKPaymentMethod.fromJson(map['paymentMethod'] as Map),
      transactionIdentifier: map['transactionIdentifier'] as String,
      paymentData: map['paymentData'] as String,
    );
  }
}

class PKPayment {
  final PKPaymentToken token;
  final PKContact? billingContact;
  final PKContact? shippingContact;
  final PKShippingMethod? shippingMethod;

  const PKPayment({required this.token, this.billingContact, this.shippingContact, this.shippingMethod});

  static PKPayment fromJson(Map map) {
    return PKPayment(
      token: PKPaymentToken.fromJson(map['token'] as Map),
      billingContact: map['billingContact'] != null ? PKContact.fromJson(map['billingContact'] as Map) : null,
      shippingContact: map['shippingContact'] != null ? PKContact.fromJson(map['shippingContact'] as Map) : null,
      shippingMethod: map['shippingMethod'] != null ? PKShippingMethod.fromJson(map['shippingMethod'] as Map) : null,
    );
  }
}

class PKPaymentError {
  final String errorType;
  final String? localizedDescription;
  final String? postalAddressKey;
  final PKContactField? contactField;

  const PKPaymentError._({
    required this.errorType,
    this.localizedDescription,
    this.postalAddressKey,
    this.contactField,
  });

  Map toJson() => {
        'errorType': errorType,
        'localizedDescription': localizedDescription,
        'postalAddressKey': postalAddressKey,
        'contactField': contactField?.toJson(),
      };

  factory PKPaymentError.contactInvalid({required PKContactField contactField, String? localizedDescription}) {
    return PKPaymentError._(
      errorType: 'paymentErrorContactInvalid',
      contactField: contactField,
      localizedDescription: localizedDescription,
    );
  }

  factory PKPaymentError.shippingAddressInvalid({String postalAddressKey = '', String? localizedDescription}) {
    return PKPaymentError._(
      errorType: 'paymentErrorShippingAddressInvalid',
      postalAddressKey: postalAddressKey,
      localizedDescription: localizedDescription,
    );
  }

  factory PKPaymentError.billingAddressInvalid({String postalAddressKey = '', String? localizedDescription}) {
    return PKPaymentError._(
      errorType: 'paymentErrorBillingAddressInvalid',
      postalAddressKey: postalAddressKey,
      localizedDescription: localizedDescription,
    );
  }

  factory PKPaymentError.shippingAddressUnserviceable({String postalAddressKey = '', String? localizedDescription}) {
    return PKPaymentError._(
      errorType: 'paymentErrorShippingAddressUnserviceable',
      localizedDescription: localizedDescription,
    );
  }
}

class PKPaymentAuthorizationResult {
  final PKPaymentAuthStatus status;
  final List<PKPaymentError>? errors;

  const PKPaymentAuthorizationResult({required this.status, this.errors});

  Map toJson() => {
        'status': status.toJson(),
        'errors': errors?.map((e) => e.toJson()).toList(),
      };
}

class PKPaymentRequestShippingMethodUpdate {
  final PKPaymentAuthStatus status;
  final List<PKPaymentSummaryItem> paymentSummaryItems;

  const PKPaymentRequestShippingMethodUpdate({required this.status, required this.paymentSummaryItems});

  Map toJson() {
    return {
      'status': status.toJson(),
      'paymentSummaryItems': paymentSummaryItems.map((e) => e.toJson()).toList(),
    };
  }
}

class PKPaymentRequestMerchantSessionUpdate {
  final PKPaymentAuthStatus status;
  final Map dictionary;

  const PKPaymentRequestMerchantSessionUpdate({required this.status, this.dictionary = const {}});
}

class PKPaymentRequestShippingContactUpdate {
  final PKPaymentAuthStatus status;
  final List<PKPaymentSummaryItem> paymentSummaryItems;
  final List<PKShippingMethod> shippingMethods;
  final List<PKPaymentError>? errors;

  const PKPaymentRequestShippingContactUpdate({
    required this.status,
    required this.paymentSummaryItems,
    required this.shippingMethods,
    this.errors,
  });

  Map toJson() {
    return {
      'status': status.toJson(),
      'paymentSummaryItems': paymentSummaryItems.map((e) => e.toJson()).toList(),
      'shippingMethods': shippingMethods.map((e) => e.toJson()).toList(),
      'errors': errors?.map((e) => e.toJson()).toList(),
    };
  }
}

class PKPaymentRequestPaymentMethodUpdate {
  final PKPaymentAuthStatus status;
  final List<PKPaymentSummaryItem> paymentSummaryItems;
  final List<PKPaymentError>? errors;

  const PKPaymentRequestPaymentMethodUpdate({
    required this.status,
    required this.paymentSummaryItems,
    this.errors,
  });

  Map toJson() {
    return {
      'status': status.toJson(),
      'paymentSummaryItems': paymentSummaryItems.map((e) => e.toJson()).toList(),
      'errors': errors?.map((e) => e.toJson()).toList(),
    };
  }
}
